### ISSUE TITLE
PyTorch ONNX export

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1333]
This issue is to track any aspects and issues on PyTorch (#1120) ONNX export.

* [x] Working script for conversion (`export_to_onnx.py`)
* [x] Fix issue with convolution
* [x] Rename script to `torch_export_to_onnx.py`
* [x] Check `model_outputs`.
* [x] Use `model_outputs` dims when not specified in `mark_as_output` (esp in case of PT).
* [x] The input names for sizes should be better (currently they are like "classes:size1" but should be like "data:size1"). (Fixed via #1362.)
* [x] The batch dim (`data:size0`), or any scalar dyn_size_ext, should not be an input, as it is redundant. (Fixed via #1362.)
* [x] `model_outputs`: Ignore dim tag equality. Check static dims, dyn dims information
* [x] Out seq lens are actually relevant, and should not be filled with dummy values? Clarify. (https://github.com/rwth-i6/returnn/issues/1333#issuecomment-1625179457) (Fixed via #1362.)
* [x] Working demo-rf + test case
* [x] Test case: Also perform ONNX inference (Fixed via #1362.)
* [ ] RF Conformer works
* [ ] Some real-world pure PT model works
* [ ] Running the resulting ONNX model in RASR
* [x] Writing a sisyphus job to run ONNX export in i6_core. (Fixed via https://github.com/rwth-i6/i6_core/pull/429.)

Other things which are less important for now:

* Avoiding `TracerWarning`s for some sanity checking code: I think it's not really possible in an easy way.

**Comment by albertz** [comment]
The initial version is done already (`export_to_onnx.py`) and works for pure PT code, so I think the main work is done. We can already close the issue. But let's anyway add further aspects on this here.

**Comment by albertz** [comment]
One aspect is that convolution with `padding="same"` is not supported by ONNX. However, we can simply work around that by doing the padding manually. The code is already there for the RF, and just needs to be used. We can use `torch.onnx.is_in_export_mode()` to test whether we should use it.

**Comment by Icemole** [comment]
As a form of summary: the tool is mainly done, but there are some [operators that aren't yet supported](https://pytorch.org/docs/stable/onnx_supported_aten_ops.html), one of which is the convolutional layer padding when specified as a string (`"same"` or `"valid"`). This prevents the demos from being run with the tool, which is one of my top priorities.

I found that it's not actually `padding="same"` which is not implemented, it's actually `type(padding) is str` what fails: a string can't be provided as `padding`, since I tried `padding="valid"` (what we had already discussed) and it failed with the same message, even when `padding=0` worked fine.

I've been investigating this issue and I'm currently trying to fix it. There's some people online having the same issues and they also provide implementations of custom padding, so it shouldn't be very hard to fix. I'll try to push a version with my changes before the weekend.

**Comment by albertz** [comment]
`padding="valid"` is just the same as `padding=0`, or not?

**Comment by Icemole** [comment]
Yes it is, but it seems that internally any padding given as a string triggers some parameter called `aten::_convolution_mode`, which isn't implemented in the ONNX conversion.

Anyway, `padding="valid"` is the easy case to fix, since then we only have to set `padding = 0`, as you said. `padding="same"` is a little more tricky in my view: it involves not only calculating the padding for each dimension (which is the easy part) but also ordering it properly in the result.

Also, padding is given as a tuple of maximum 3 dimensions `([[depth,] height,] width)` depending on the dimensions of the convolutional layer (3D, 2D or 1D), so I have to investigate a bit the internals in order to check if there's redundant padding inserted that we don't desire: for instance, with a kernel size of 2x2 and a stride of 1, I think we'd only need to insert an additional row and column, but maybe this tuple considers symmetric padding and inserts one row at the start and another at the end, and the same for the column, making the output size not match the input size.

**Comment by albertz** [comment]
> Anyway, `padding="valid"` is the easy case to fix, since then we only have to set `padding = 0`, as you said. 

Yes, my question was just to confirm this. So we can just always do this then, no matter if ONNX or so. Just:
```python
if padding == "valid":
    padding = 0
```

> `padding="same"` is a little more tricky in my view: it involves not only calculating the padding for each dimension (which is the easy part) but also ordering it properly in the result.

We already have the code for adding the padding manual. See the code. It basically does:
```python
if padding == "same" and <manual_padding_required>:
    ...  # add padding to input
    padding = "valid"
```

Nothing else needs to be done, or not?

**Comment by albertz** [comment]
Btw, next step after the current demo works: Test our RF Conformer implementation.

**Comment by albertz** [comment]
Actually, I just wonder now: Instead of manually adding the padding to the input, could we maybe just set `padding = (filter_size - 1) // 2`? But probably not really in all cases. E.g. only works with odd filter_size. No striding. Etc. But it might be worth to still have this case, as it might be faster than the manual padding.

**Comment by Icemole** [comment]
I was able to insert `torch.onnx.is_in_export_mode()` into the already developed code which creates manual striding. However, this only works for the RF demo: the PT demo as it is can't be executed, since it also uses `padding="same"` but we have no handler for pure PT code. Should we just let these details to the user, or also map `"same"` to a valid padding size?

**Comment by albertz** [comment]
> I was able to insert `torch.onnx.is_in_export_mode()` into the already developed code which creates manual striding. However, this only works for the RF demo: the PT demo as it is can't be executed, since it also uses `padding="same"` but we have no handler for pure PT code. Should we just let these details to the user, or also map `"same"` to a valid padding size?

Change `padding="same"` to `padding=2` in the Torch demo. That should fix this issue.

**Comment by albertz** [comment]
> Internally, the `forward()` implementation of convolutional layers is as follows ...

This is only for the case `padding_mode != 'zeros'`, which we don't support at all currently, right?

**Comment by albertz** [comment]
I'm not exactly sure what you mean. The padding numbers don't need to be calculated at all. `F.pad` will just add them. `F.pad` will pad the input. The input is different for every `forward()` call of course, so there is nothing we can cache.

**Comment by Icemole** [comment]
Oh I see, yes, I neglected the effect of the dynamic dimensions on convolution, I was thinking more like image processing where the input is usually the same. Nevermind.

**Comment by Icemole** [comment]
I've pushed some commits that should fix some of the issues with the exporting tool. Hopefully I explained everything nicely.

With the current state of the tool (4785332), I can properly run both configs by the ONNX exporting tool. However, the graphs differ greatly. The number of operations in the RF graph is much bigger than the number of operations in the PT graph. At first sight, the PT graph looks much cleaner than the RF graph. I haven't studied the differences yet, I'll look at it on Monday. In the meantime, you can find both graphs below.

[onnx_pt_graph.txt](https://github.com/rwth-i6/returnn/files/11518450/onnx_pt_graph.txt)
[onnx_rf_graph.txt](https://github.com/rwth-i6/returnn/files/11518451/onnx_rf_graph.txt)

**Comment by Icemole** [comment]
I tried exporting again the RF demo model with the current changes, and the graph is much smaller now! Basically almost equal to the PyTorch graph: [onnx_rf_graph_new.txt](https://github.com/rwth-i6/returnn/files/11528190/onnx_rf_graph_new.txt) :partying_face:

**Comment by albertz** [comment]
> I tried exporting again the RF demo model with the current changes, and the graph is much smaller now! Basically almost equal to the PyTorch graph: [onnx_rf_graph_new.txt](https://github.com/rwth-i6/returnn/files/11528190/onnx_rf_graph_new.txt) 🥳

The actual computation is not just almost identical, but exactly identical, as far as I can see it.

In addition, it defines the output sizes. Which are just the same as the input sizes, because of padding "same", so it's just an identity function here.

I think when there are new calculations of dims, this will still be unnecessary complex. I think we can improve this more. But not sure how important this is.

**Comment by albertz** [comment]
I think I reopen this to keep track of the remaining issues.

One remaining issue now is that we do not really check `model_outputs`, whether that matches what the user actually returns in the forward-step function. And in case of the PT demo, where the user does not provide the dims in the `mark_as_output` call, it does not exactly match the dims from `model_outputs`.

**Comment by Icemole** [comment]
There seems to be one last thing to figure out before the tool is fully operational. I get an error while running the PyTorch demo:

```
RuntimeError: number of output names provided (3) exceeded number of outputs (1)
```
This is the [full stack_trace](https://github.com/rwth-i6/returnn/files/11528256/stack_trace.txt).

I think this is happening because somehow the RF model is outputting the output sizes as well, while the PT model isn't.

PT and RF don't exactly behave the same when calling `rf.get_run_ctx().outputs.as_raw_tensor_dict()`. PT instantiates the dimensions of `rf.get_run_ctx().outputs.data` to actual dimensions, while RF keeps the batch dim and the time dim as dynamic dimensions. Not sure if this change in behavior is intended.

Since in the PT demo we don't have access to batch/time dims, the output dimensions are instantiated to actual values. However, this isn't the case in the RF demo, where we have access to the shape of the original tensor as `Dim` classes.
Since these dimensions are also marked as output, this is the essential difference.

I introduced a workaround in 47853321, which is why I was able to post the PT graph above, but it was reverted on bb40b4ca. Is there any other way to fix this?

**Comment by albertz** [comment]
I would maybe extend `init_forward_step_run_ctx` to pass a `model_outputs` template. Then in `mark_as_output`, it would check for that and follow the logic I described above. Also in the end, some function `check_outputs_match` or so, which checks whether the all outputs are given.

**Comment by albertz** [comment]
Another thing: This script `export_to_onnx.py`, this is actually PT specific, right? We should then rename it, to `torch_export_to_onnx.py` or so.

**Comment by Icemole** [comment]
> some function check_outputs_match or so, which checks whether the all outputs are given

And if they don't match, which is what we're seeing here, what to do? We've detected the error, but should we fix it, or do we leave the user to it? If we fix it, do we give priority to the expected outputs, or to the actual outputs? I would argue that the actual outputs would have more importance, but it's an "unexpected" situation in which maybe an automatic fix isn't the most complete thing to do...

**Comment by Icemole** [comment]
> This script export_to_onnx.py, this is actually PT specific, right? We should then rename it, to torch_export_to_onnx.py or so.

Yes, it works for the PT backend, independently of whether the RF frontend or actual PT code is being used. I'm fine with the name change.

**Comment by Icemole** [comment]
There's one more item which I would probably add to the list of items to check and that we briefly discussed: running the resulting ONNX model in RASR.

**Comment by albertz** [comment]
> [3c22503](https://github.com/rwth-i6/returnn/commit/3c22503434305f1f0593317ed0e077cf1420b3b2) addresses "Check `model_outputs`" in the main checklist.

There were some problems with that. See my updated commit.

In this commit, I also already overtake the dims in case they are not specified, so this should be implemented as well now.

**Comment by albertz** [comment]
I don't exactly understand the problem. The dimensions should already be filled, or not? We already called `tensor_dict_fill_random_numpy` before that? So why are they not filled?

**Comment by Icemole** [comment]
The dimensions of `extern_data` were filled, but not the dimensions of `model_outputs`.

**Comment by albertz** [comment]
> The dimensions of `extern_data` were filled, but not the dimensions of `model_outputs`.

Your error was about the batch dim, which should be the same?

> And as I mentioned in my previous comment, I'm not sure we can infer any other dimension from `extern_data` which isn't the batch dimension,

But the error is about the batch dim?

> because we aren't sure that these will be kept throughout the computation, as it happens for instance with a convolutional layer with `padding="valid"`.

Yes sure, those will be different. I think in `convert_to_tensor` we could just define it if it is undefined.

**Comment by Icemole** [comment]
> But the error is about the batch dim?

Yes, but it only crashed in the batch dim because the batch dim is the first dim to be checked. The time dim doesn't have any actual value either, so if we somehow fix the batch dimension, the code will also crash when checking the actual value of the time dim. Actually, any dimension other than the input/output dimension will not have any value if defined similarly to what we do in the `demo-torch.config`.

> I think in convert_to_tensor we could just define it if it is undefined.

Which values would it have? Arbitrarily filled, as in `tensor_dict_fill_random_numpy`? If so, it would probably fix the error I'm reporting but not the main issue, which is that we don't know what the output dim values should be before we run the model. With that change, I think it would probably fail in `mark_as_output()`, when asserting `expected_output.dims == tensor.dims`.

**Comment by albertz** [comment]
In `convert_to_tensor`, you have the actual value of the shape. Just not individual seq lens, but that's ok, just set dyn_size_ext to Tensor(dims=[], raw_tensor=shape[...]).

**Comment by Icemole** [comment]
I was able to code a solution locally, but I think I might have overengineered it. I'd like to know your thoughts before pushing it.

After [this line of `torch/frontend/_backend.py`](https://github.com/rwth-i6/returnn/blob/master/returnn/torch/frontend/_backend.py#L657), I added the following code (note that I add the `assert` and the `return` that are currently in the file for reference):

```
assert isinstance(value, torch.Tensor)
# Fill dynamic dims if we have enough information about them
for i, dim in enumerate(dims):
    if dim.is_dynamic() or dim.is_batch_dim():
        if not dim.dyn_size_ext:
            # Batch dim for now only until we find why the value is not being copied over
            raw_tensor = torch.tensor(value.size(i), dtype=torch.int32)
            dim.dyn_size_ext = Tensor("batch", dims=[], raw_tensor=raw_tensor, dtype="int32")
        elif not dim.dyn_size_ext.raw_tensor:
            raw_tensor = torch.ones(
                len(dim.dyn_size_ext.dims), dtype=TorchBackend.as_dtype_raw(dim.dyn_size_ext.dtype)
            )
            raw_tensor[-1] = value.size(i)
            dim.dyn_size_ext.raw_tensor = raw_tensor
return Tensor(name, dims=dims, dtype=dtype, sparse_dim=sparse_dim, raw_tensor=value)
```

I placed this code in `torch/frontend/_backend.py` because I couldn't figure out a way to get the raw tensor dimension sizes in the frontend, independently of the backend. If you know that such a thing is already implemented, or you think it would be a good addition, let me know.

The code would be greatly simplified when we find what's happening with the batch dimension (i.e. why it isn't the same object, see my comment above), as the first `if` would be removed altogether. I only put it so that I could have a locally working tool.

I found that, for the dynamic dimensions (at least for the time dimension in the PyTorch demo), everything was set except for the raw tensor, which made it easier to handle. Therefore, I just focused on properly filling the `dim.dyn_size_ext.raw_tensor` attribute (see the `elif` part).

I see two main drawbacks with this solution:

1. If we want to be consistent, it will need to be implemented for both the PyTorch backend, as well as for the TensorFlow backend. It wouldn't be strictly needed to implement it in the Tensorflow backend though, as it's only for exporting PyTorch models.
2. [The resulting ONNX graph for the PyTorch demo](https://github.com/rwth-i6/returnn/files/11567177/onnx_pt_graph.txt) is bigger than expected. Some comments in the resulting ONNX graph report point to the fact that line 660 could be the culprit, which contains the conditional checks of `if dim.is_dynamic() or dim.is_batch_dim()`. This might be solved if we fix whatever's happening with the batch dimension that I mention above, but I'm not sure if the graph will be as short as the RF one as we introduce additional tensor creation code for handling this situation.

By the way, the RF graph isn't modified by these changes.

**Comment by albertz** [comment]
Remove the code for the batch dim. We should not add random workaround code just because we are too lazy to debug and fix a problem, or better understand what the problem is. Or remove at least the comment because it's just wrong. The code is correct, or at least valid to do.

I don't understand your code branch on `elif not dim.dyn_size_ext.raw_tensor:`. Why do you have that? The code is in any case wrong. You should not put some dummy values in this case. That does not make sense.

> I couldn't figure out a way to get the raw tensor dimension sizes in the frontend

Well you must access the backend for this, obviously. So code like this:
```python
value_backend = get_backend_by_raw_tensor_type(type(value))
value_backend.get_shape_tuple_raw(value)[axis]
```
Or so.

You actually see similar code already in `convert_to_tensor`.

You should not put this code into the backend. It should be in the frontend. Just like there is similar code already in `convert_to_tensor`.

> I found that, for the dynamic dimensions (at least for the time dimension in the PyTorch demo), everything was set except for the raw tensor, which made it easier to handle. 

What do you mean? You mean `dyn_size_ext` was set? But that would be wrong. Why is it set? It should not be set. Actually this makes it problematic. Anything else than setting `dyn_size_ext` by us here does not make sense.

> Therefore, I just focused on properly filling the `dim.dyn_size_ext.raw_tensor` attribute (see the `elif` part).

As said above, this does not make sense. You cannot set this. We cannot know what should get in there if this is not a scalar.

**Comment by Icemole** [comment]
Thanks for the feedback, `get_shape_tuple_raw()` seems to be exactly what I need. I will change the code as you proposed.

> You mean dyn_size_ext was set? But that would be wrong. Why is it set?

A little debugging shows that it's set in `returnn/tensor/_tensor_extra.py`, in the function `infer_dim_tags()`, which calls `_infer_dim_tags_tuple_from_shape()`. More specifically, [in this line](https://github.com/rwth-i6/returnn/blob/master/returnn/tensor/_tensor_extra.py#L3656), which covers the case where the inferred `Dim` object doesn't have a `dyn_size_ext` attribute, and if this is the case then it sets `dyn_size_ext`. However, it didn't set the raw tensor property, most probably because there's no information about its size.

> It should be sth like if not dim.is_dim_known(). You want to check if it is undefined.

If we leave the code like this (i.e. we don't touch the `_infer_dim_tags_tuple_from_shape()`), then the call to `dim.is_dim_known()` will return `True` as [in `is_dim_known()` there's a specific branch for a `Tensor` that has some `dyn_size_ext` available, independently of its value](https://github.com/rwth-i6/returnn/blob/master/returnn/tensor/_dim_extra.py#L759).

**Comment by albertz** [comment]
> > You mean dyn_size_ext was set? But that would be wrong. Why is it set?
> 
> A little debugging shows that it's set in `returnn/tensor/_tensor_extra.py`, in the function `infer_dim_tags()`, which calls `_infer_dim_tags_tuple_from_shape()`. More specifically, [in this line](https://github.com/rwth-i6/returnn/blob/master/returnn/tensor/_tensor_extra.py#L3656), which covers the case where the inferred `Dim` object doesn't have a `dyn_size_ext` attribute, and if this is the case then it sets `dyn_size_ext`. However, it didn't set the raw tensor property, most probably because there's no information about its size.

Ah yeah I should have known this. This is somewhat legacy logic. This is reasonable for extern_data in TF where it will be set to the placeholder then. It was also a reasonable default assumption that we would have such a shape [B] for the seq lens.

We should just fix this. I.e. either we force the user to not use the legacy logic for this case. Or we just do some short post-processing of it and remove and `dyn_size_ext` which has no raw_tensor set. Maybe this latter option is nicer for the user. But we should be careful. We should not reset dims which also occur in `extern_data`, because in `extern_data`, we still want this behavior, and `dyn_size_ext` should not be reset.

> > It should be sth like if not dim.is_dim_known(). You want to check if it is undefined.
> 
> If we leave the code like this (i.e. we don't touch the `_infer_dim_tags_tuple_from_shape()`), then the call to `dim.is_dim_known()` will return `True` as [in `is_dim_known()` there's a specific branch for a `Tensor` that has some `dyn_size_ext` available, independently of its value](https://github.com/rwth-i6/returnn/blob/master/returnn/tensor/_dim_extra.py#L759).

As I said, we should not need to deal with the case that there is `dyn_size_ext` available. At least that was what I thought initially is most reasonable.

We should think a bit more in principle, what do we actually want, what would we expect, what behavior is different from the expected behavior, is there something we should maybe fix or change?

I was also thinking more about the `model_outputs`. There is one problem actually: We cannot really know all the dims of the output in advance. E.g. the output time dimension might be of some arbitrary shape. In some cases we might still compute it in advance, but not always, e.g. when it is totally dynamic. So the whole idea that we can define the model outputs this way is somewhat flawed. Or at least this check in `mark_as_output` does not really make sense:
```python
assert expected_output.dims == tensor.dims
```

How would we fix that? Not use `model_outputs` at all? But I think sth like `model_outputs` is nice to have to define the output properly, and maybe even necessary.

Maybe a somewhat more custom check for the dim:
```python
assert len(expected_output.dims) == len(tensor.dims)
for expected_dim, tensor_dim in zip(expected_output.dims, tensor.dims):
  if not expected_dim.is_dim_known():
    # If tensor dim is dynamic, we might not have known the dim beforehand,
    # so this is reasonable.
    # (Or maybe even remove this check and allow even static tensor dim here?)
    assert tensor_dim.is_dynamic()
    # (We might maybe want to replace the dim in expected_output?)
    # (Or we maybe might want to use expected_dim.declare_same_as(tensor_dim)?)
  else:
    assert expected_dim == tensor_dim
```

**Comment by Icemole** [comment]
Might I ask why is it troublesome that `dyn_size_ext` is set with everything but the raw tensor property? This is enacted when the size of the dimension is unknown when creating the tensor with the corresponding dimensions, which I believe will be very common in everything which isn't an RF setup. Even if it was initially meant as a workaround to the TF code, I think if we know how to handle it, it could be beneficial in this situation.

I definitely agree in the fact that we can't estimate it beforehand: estimating it would be equivalent to checking the dimensions of the output after a forward call, which is exactly what we're trying to do in the first place. Still, I think we agree in the fact that `model_outputs` is a nice property to have, so we should probably make it work.

**Comment by albertz** [comment]
> Might I ask why is it troublesome that `dyn_size_ext` is set with everything but the raw tensor property?

Because you don't know what the correct `dyn_size_ext` should be in advance. It is set just because of a heuristic, an assumption, but this assumption might be wrong. We should not set it when it is unknown. In fact, this assumption is actually wrong in our case. It assumes [B] shape, but in case of PyTorch, we would get [] shape, a scalar, for the `dyn_size_ext`.

> if we know how to handle it

How to handle what?

Let's leave current heuristics and assumptions away from this discussion for now. Let's say there are no heuristics and assumptions at all. The user completely specifies everything. If you use `out_time_dim = Dim(None, name="out-time")`, this is an undefined dim (`is_dim_known()` returns False, `dyn_size_ext` is not set). So the user could use that. Or the user could specify more explicitly that he expects a [B] shape `dyn_size_ext` in the output. Then the user could specify that.

My argument was, whatever the user has specified, we should follow this.

My other argument was, we cannot have equality of dims in all cases. Esp in the case when the dim cannot be known in advance. In that case, the user would specify an undefined dim. But we would not have equality of the undefined dim to what we get in the end. So we cannot do this check `assert expected_output.dims == tensor.dims`. We would treat undefined dims specially.

Btw, even with pure PyTorch, we could say if the user specifies `model_outputs`, with a defined `dyn_size_ext` (except raw_tensors), then we require that the user also uses the same dim in `mark_as_output`, and also that the user explicitly defines raw_tensor of that dim. So this is up to the user.

We could also say, if the user specifies `dyn_size_ext` with given scalar shape ([]), then we can infer it automatically, but if the user specifies an undefined dim, then we don't infer it automatically (in contrast to what I said before) but simply ignore this dim.

So, we should first decide on what behavior do we want, what would be the meaning of each possible user specification:
1. undefined dim
2. defined static dim
3. defined dynamic dim, scalar `dyn_size_ext`
4. defined dynamic dim, non-scalar `dyn_size_ext`

Once we have decided on that, we can get back to the question whether any heuristics or implicit assumptions make sense, or should be disallowed, or changed.

**Comment by albertz** [comment]
Some suggestion for possible behavior for each of those cases:

1. undefined dim:

    - `mark_as_output` would simply ignore this.

2. defined static dim:

    - `mark_as_output` requires dim equality.

3. defined dynamic dim, scalar `dyn_size_ext`:

    - `mark_as_output` requires dim equality.
    - If `dyn_size_ext.raw_tensor` is unset here, we can automatically set it via the shape.

4. defined dynamic dim, non-scalar `dyn_size_ext`:

    - `mark_as_output` requires dim equality.
    - Require that the user has defined `dyn_size_ext.raw_tensor` before the `mark_as_output` call.

**Comment by Icemole** [comment]
I tried to put into code the last comment of yours. Feel free to fix it whenever applicable.

With respect to the third point, `fill_raw()` isn't implemented in the torch backend. It shouldn't be hard to implement, but I don't have more time. Any other alternative that creates a simple raw tensor is welcome.

**Comment by albertz** [comment]
Ok, I thought we would discuss this first before actually implementing this. Because I'm not really sure if this si the right way. Also, there are a couple of open questions.

E.g. the use of `declare_same_as`: In my last comment, I actually did not use this at all. And I would also want to avoid this. See #975. In my last comment, why would you need it there? It would be wrong to use it, if you followed my last comment.

Also, if we decided what we want to have (still before implementing it actually, before doing any changes to the code), there is still the big open question if any of the current heuristics are maybe bad and should be changed somehow.

**Comment by albertz** [comment]
I cleaned up your code. But anyway, I think we still should actually think about, and discuss whether this suggested logic makes sense or not. Or did you basically agree to it, and that's why you implemented it?

**Comment by Icemole** [comment]
> I would also want to avoid [declare_same_as]

I see, I thought "overtaking" the dim meant using `declare_same_as`.

> Or did you basically agree to it

Yes, I had agreed to it. I understand the idea of not being able to directly check that the expected dimensions are the same as the actual dimensions, since the dynamic dimensions might not have sizes. And anyway this proposition only checks for values in the dimensions, and overwrites the values of the expected dimensions whenever the actual dimensions can provide information that the expected dimensions don't have yet. Moreover, your code is even laxer by not using `declare_same_as` and removing some assertions. I think this is one of the laziest schemes we can get right now, which is what we wanted from the beginning.

**Comment by albertz** [comment]
Ok, if we then decided on the behavior, this then gets us back to the question, are the current heuristics reasonable? Probably not. At least when the user specifies this:
```python
model_outputs = {
  "output": {"dim": 5}
}
```
Via the heuristics, it would create a tensor with shape `[batch_dim, time_dim, feat_dim]`, with `feat_dim = Dim(5)`, `time_dim = Dim(Tensor([batch_dim]))`. But this is not what we want, at least for the `time_dim`. Or maybe it would just be unexpected. This would be the case "defined dynamic dim, non-scalar dyn_size_ext". However, likely the user has not specified `dyn_size_ext.raw_tensor` before the `mark_as_output` call, so it fails.

So, how to solve this? Change the heuristics to not create `time_dim` this way, but instead to simply create `time_dim = Dim(None)`, i.e. an undefined dim? Or just require that the user specifies it correctly/explicitly? I.e. like this:
```python
model_outputs = {
  "output": {
    "dims": [batch_dim, Dim(None, name="time"), Dim(5, name="feature")],
    "dtype": "float32",
  },
}
```

**Comment by albertz** [comment]
Small change after discussion: We ignore dim tag equality in all cases in `model_outputs`. We just check static dims, and take the information which dims are dynamic.

**Comment by albertz** [comment]
More like this:

In no cases, we check dim equality.

1. undefined dim:

    - `mark_as_output` assumes dynamic dim.

2. defined static dim:

    - `mark_as_output` requires same static dim (but not dim tag equality).

3. defined dynamic dim:

    - `mark_as_output` assumes dynamic dim.

**Comment by albertz** [comment]
Btw, for the discussion on how to set `model_outputs`, or whether the current heuristics are ok: Actually, they are ok for this behavior, so no need to change anything about that.

**Comment by Icemole** [comment]
I'll reopen this issue since we still have some things left to be done in the checklist.

**Comment by albertz** [comment]
Regarding the batch dim tag recreation issue (please post the other information on that), I'm playing around with that here: https://github.com/rwth-i6/returnn/pull/1343

**Comment by Icemole** [comment]
Batch dim tag recreation issue: I tracked down [the issue that I mentioned a while ago](https://github.com/rwth-i6/returnn/issues/1333#issuecomment-1561494324), the one that led to the rework of the logic (which didn't solve the original issue). The stack trace is given in my original comment.

After debugging, I found out that the batch dimension and time dimension were the same dimension (both through the `==` operator and through the `is` operator) when executing the ONNX exporting tool with the RF demo, but they were NOT the same when executing the tool with the torch demo.

While further debugging, I found that the batch dimension was being recreated [here](https://github.com/rwth-i6/returnn/blob/master/returnn/tensor/_tensor_extra.py#L3577).

I attach two captures corresponding to the debugger. I placed a checkpoint on line 150 of the torch ONNX exporting tool, but it can be set wherever, as long as both `extern_data` and `model_outputs` are defined (line 150 is the line after setting `model_outputs`). The first capture corresponds to the state of the batch dimension in `extern_data`, which as you can see contains an actual dimension because it was already filled by the tool.

![](https://github.com/rwth-i6/returnn/assets/31628502/92890d75-cf47-41b6-8095-acc0060e9c3b)

The second capture corresponds to the state of the batch dimension in `model_outputs`, which is yet undefined.

![](https://github.com/rwth-i6/returnn/assets/31628502/c790ec68-eba4-496c-a0df-8a0a9dcc431e)

**Comment by albertz** [comment]
Ok, #1343 is merged now, which should fix the issue on batch dim recreation. Can you test?

Btw, it would really be helpful to have these test cases as discussed. Then we can more easily see if all works fine or when sth breaks.

**Comment by Icemole** [comment]
Yes, this fixes the issue on batch dim recreation. Now there's another issue:
```
RuntimeError: number of output names provided (3) exceeded number of outputs (2)
```
I haven't had time to look at the error, but in this thread I also reported a similar error: https://github.com/rwth-i6/returnn/issues/1333#issuecomment-1556693295.

See [full stack trace](https://github.com/rwth-i6/returnn/files/11771749/stack_trace.txt).

With respect to test sets, I definitely agree, they would be a very welcome addition, but first I'd like to leave everything finished. I haven't had almost any time to look at the ONNX exporting tool though...

**Comment by albertz** [comment]
We have `output_names = ['output', 'output:size0', 'output:size1']` here.

Can you make a breakpoint in `forward_step` / `mark_as_output`, and look at what gets added to the outputs? Or directly in `rf.get_run_ctx().outputs.as_raw_tensor_dict()`, what is being returned here? This is not consistent to the output names. Why is that?

**Comment by albertz** [comment]
I added an extra check, which should show a better error when this goes wrong.

**Comment by Icemole** [comment]
I've been able to check why the tool is still failing for the PT demo. We patched the fact that the batch dimension didn't have a `dyn_size_ext.raw_tensor` property by using the global batch dimension, but we didn't do this for the time dimension.

Specifically, what happens in `rf.get_run_ctx().outputs.as_raw_tensor_dict()` is that the output time dimension doesn't have a `dyn_size_ext.raw_tensor` property, because we didn't give it any. Therefore, the output for the key `output:size1` (time dimension) is the value `None`, which fails to be converted into a normal key-value in the internal `_flatten` operation. I assume this `_flatten` function has some kind of detection of values labeled as `None`.

To some degree this was intended, since we agreed in the fact that the output time dimension couldn't get the `dyn_size_ext` property from the input time dimension, which is understandable because the time dimension can change in the computation. However, we need to ensure that the output time dimension has some `dyn_size_ext.raw_tensor` property. If we don't want to let the user control this, then we probably need to control it ourselves after the call to `self.forward_step_func()` and before the call to `rf.get_run_ctx().outputs.as_raw_tensor_dict()`, right?

**Comment by albertz** [comment]
Ah yes that makes sense. We simply forgot about this.

Currently in `mark_as_output`, the logic is this:
```
expected_output = self.expected_outputs.data[name] if self.expected_outputs else None
if dims is None and expected_output:  # this is our case
    dims = expected_output.dims
```
However, then we don't do any further logic with the `dims`. I think we previously already discussed that we would want to have this logic:
Automatically set the `raw_tensor`.

We actually had this code already some time ago. It was ok for the scalar case (dyn_size_ext being a scalar). However, for other cases, e.g. [B], it is a bit unclear what to do. And I think this is the case we have here, right? We could do:
```python
dyn_size_ext.raw_tensor = torch.full([d.get_dim_value() for d in dyn_size_ext.dims], tensor.shape[axis])
```
I think you even had some similar code at some earlier time, as far as I remember?
The problem with this code is that it might make debugging annoying when the user actually wants to set the dyn_size_ext correctly explicitly but forgot to do and then it would silently just being defined here as these dummy values. Maybe we should print a warning? For all non-scalar cases.

Btw, if we would have the test case, then it would make it really easier to debug this, as we can directly see the error. I think it makes sense to have the test case already now such that this can be used to track the current state and always see the current error.

**Comment by Icemole** [comment]
By the way, the `_check_matching_outputs()` call is failing, maybe in `outputs_raw_keys` you need a `as_raw_tensor_dict()` somewhere.

**Comment by albertz** [comment]
> By the way, the `_check_matching_outputs()` call is failing, **edit**: when executing the tool for both the RF demo and the PT demo. Maybe in `outputs_raw_keys` you need a `as_raw_tensor_dict()` somewhere.

For that such a test case would really be helpful.

**Comment by Icemole** [comment]
Tests added. Let me know if I should add them separately.

**Comment by Icemole** [comment]
While the tool does run correctly now for both the PT and the RF demos, I've seen the following warning for the PT demo:

```
WARNING:root:Cannot infer dynamic size for dim Dim{'time:var-unk:output'[B?]} from output 'output'
Tensor{'raw_tensor', [B?,'time:var-unk:output'[B?],F'feature:output'(2)]}.Using Tensor{'full', [B?], dtype='int32'}
as fallback.
```

Of course, this is intended as per your changes. But more worrying is the following warning when executing the demo with `torch==2.0.1`, which I believe is key to our issues:

```
======================= 0 NONE 0 NOTE 0 WARNING 0 ERROR ========================

/home/nbeneitez/.venv/returnn_pytorch/lib/python3.10/site-packages/torch/onnx/utils.py:1636: UserWarning:
The exported ONNX model failed ONNX shape inference.The model will not be executable by the ONNX Runtime.
If this is unintended and you believe there is a bug,please report an issue at https://github.com/pytorch/pytorch/issues.
Error reported by strict ONNX shape inference: [ShapeInferenceError] (op_type:ConstantOfShape, node name:
ConstantOfShape_14): input typestr: T1, has unsupported type: tensor(int32) (Triggered internally at
../torch/csrc/jit/serialization/export.cpp:1407.)
  _C._check_onnx_proto(proto)

Process finished with exit code 0
```

Note that the model is generated, but the part of `The model will not be executable by the ONNX Runtime` is concerning. Moreover, the exported graphs for version 1.13.1 and version 2.0.1 are the same.

Maybe this error was also present on `torch==1.13.1` but was never caught by torch... Is there any way to perform inference with the exported model with the torch demo config, or with an arbitrary config? If not, are there any setups in RETURNN for inference with an ONNX on which I can test the output of the tool?

**Comment by albertz** [comment]
> But more worrying is the following warning when executing the demo with `torch==2.0.1`, which I believe is key to our issues:

What do you mean by "our issues"? All issues we had so far are fixed now, or not? Is this a new issue, or some existing issue we already had seen before?

On the warning: I don't really understand the problem. What exactly does it mean that it failed ONNX shape inference? And why did it fail? And what exactly did fail?

And what does this unsupported type mean? int32 should be supported? Or is this about the shape dim, which expects int64 only or so?

You could try to add this hack in the code, just to see if the error goes away:
```python
dim.dyn_size_ext.dtype = "int64"
```
Before the line where we do `rf.cast(dim_value, dtype=dim.dyn_size_ext.dtype)`.

> Is there any way to perform inference with the exported model with the torch demo config, or with an arbitrary config?

Yes, obviously of course there is (there must be, how would you ever use the ONNX graph if you cannot use it?).

See for example here: https://github.com/rwth-i6/i6_models/blob/52410fafa72b20cee174ece96d5d850bc73c4c61/tests/test_blstm.py#L36

**Comment by Icemole** [comment]
> our issues

I meant the possible issue that might arise from the warning, with so far unknown consequences (see below for a list of consequences :smile:).

> On the warning: I don't really understand the problem. What exactly does it mean that it failed ONNX shape inference? And why did it fail? And what exactly did fail?
> And what does this unsupported type mean? int32 should be supported? Or is this about the shape dim, which expects int64 only or so?

This is probably about the shape dimension (`%classes:size0`), see below for a detailed debugging report.

---

I have tried your hack, and the error doesn't go away. Weirdly enough, the fallback's dtype is updated, but the error is still the same, pointing to `tensor(int32)`:

```
WARNING:root:Cannot infer dynamic size for dim Dim{'time:var-unk:output'[B?]} from output 'output'
Tensor{'raw_tensor', [B?,'time:var-unk:output'[B?],F'feature:output'(2)]}.Using Tensor{'full', [B?], dtype='int64'}
as fallback.
/home/nbeneitez/.venv/returnn_pytorch/lib/python3.10/site-packages/torch/onnx/utils.py:1636:
UserWarning: The exported ONNX model failed ONNX shape inference.The model will not be executable by the
ONNX Runtime.If this is unintended and you believe there is a bug,please report an issue at
https://github.com/pytorch/pytorch/issues.Error reported by strict ONNX shape inference:
[ShapeInferenceError] (op_type:ConstantOfShape, node name: ConstantOfShape_14): input typestr: T1,
has unsupported type: tensor(int32) (Triggered internally at ../torch/csrc/jit/serialization/export.cpp:1407.)
  _C._check_onnx_proto(proto)
```

Thank you for the pointer to the ONNX inference test. I've tried it, and it fails with the following error, both in torch version 1.13.1 and in version 2.0.1:

```
Traceback (most recent call last):
  File "/home/nbeneitez/Documentos/work/repos/returnn_pytorch/test_onnx_inference.py", line 10, in <module>
    session = ort.InferenceSession(onnx_model)
  File "/home/nbeneitez/.venv/returnn_pytorch/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 383, in __init__
    self._create_inference_session(providers, provider_options, disabled_optimizers)
  File "/home/nbeneitez/.venv/returnn_pytorch/lib/python3.10/site-packages/onnxruntime/capi/onnxruntime_inference_collection.py", line 424, in _create_inference_session
    sess = C.InferenceSession(session_options, self._model_path, True, self._read_config_from_model)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidGraph: [ONNXRuntimeError] : 10 : INVALID_GRAPH : Load model from /tmp/nbeneitez/returnn/demos/demo-torch/model.005.onnx failed:This is an invalid model. Type Error: Type 'tensor(int32)' of input parameter (onnx::ConstantOfShape_25) of operator (ConstantOfShape) in node (ConstantOfShape_14) is invalid.

Process finished with exit code 1
```

This proves that, while the error was correctly printed in 2.0.1, it was concealed in 1.13.1.

The error lists that the failing operator is used in the input of `onnx::ConstantOfShape_25`. Such operator is declared in the graph as:

```
  %onnx::Unsqueeze_23 : Long(1, strides=[1], device=cpu) = onnx::Constant[value={0}]()
  %onnx::Concat_24 : Int(1, strides=[1], device=cpu) = onnx::Unsqueeze(%classes:size0, %onnx::Unsqueeze_23)
  %onnx::ConstantOfShape_25 : Int(1, strides=[1], device=cpu) = onnx::Concat[axis=0](%onnx::Concat_24) # /home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:672:0
```

I'm using ONNX opset 19 (ONNX v1.14.0).

---

The difference between the torch and the RF graphs is that the torch graph has many more operations to calculate the `%output:size1` operator, while the RF graph has simply the following operator:

```
  %output:size1 : Int(*, strides=[1], requires_grad=0, device=cpu) = onnx::Identity(%classes:size1)
```

I found that this is probably the case because the `%classes:size1` input isn't provided in the list of inputs. Check out the [exported graph which I shared two comments ago](https://gist.github.com/Icemole/8bf3d81d0430d83afc32d57f1604df09). However, debugging shows that the input parameters (that is, `extern_data.keys()` show all the values, including `classes:size1`: `['data', 'data:size0', 'data:size1', 'classes', 'classes:size0', 'classes:size1']`. This should be reproducible in your setup, since I didn't change anything from the latest commits (except adding your hack).

---

While debugging, I saw that after the call to `rf.full()` in the fallback you added to `RunCtx._output_tensor_from_raw`, returns the size of the padded tensor, and not the individual tensor sizes. Could this pose any issue later on?

**Comment by albertz** [comment]
I think the actual failing op in the ONNX graph is this:
```
  %onnx::Add_26 : Int(*, device=cpu) = onnx::ConstantOfShape[value={0}](%onnx::ConstantOfShape_25) # /home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:673:0
```
As I understand the format, it is basically:
```
 %<NodeName> : <Type>(<value?>, <opts?>) = onnx::<OpType>[<attribs?>](<inputs>)  # <Python-code-line>
```
The node names might be renamed automatically according to some logic which I don't exactly understand?
However, the node types is what better identifies the op. So when it mentions `operator (ConstantOfShape)`, this is the only node in the graph with this op type.

**Comment by albertz** [comment]
To extend my prev comment, as I understand the error now, I think it complains that in `torch.full`, the `shape` we provide is of type int32.

Basically, it would mean, this standalone code would fail:
```python
out = torch.full([torch.tensor(3, dtype=torch.int32)], 0)
```
Can you test this?

If this is the case, this is maybe a bug in PyTorch? Because that code is valid PyTorch code, so I think it should be able to export it. In that case, can you make a bug report, and link it here?

**Comment by albertz** [comment]
I pushed a small change which makes sure that the `shape` is int64 for this `torch.full` in case of ONNX exporting. Can you try whether that fixes it?

In any case, verify the standalone `torch.full` code in my prev comment, and then maybe make a PyTorch bug report about this.

**Comment by Icemole** [comment]
> `out = torch.full([torch.tensor(3, dtype=torch.int32)], 0)`

This standalone code doesn't fail.

---

I got the following error while running your fix:

```
TypeError: full() received an invalid combination of arguments - got (Tensor, Tensor, dtype=torch.dtype), but expected one of:
 * (tuple of ints size, Number fill_value, *, tuple of names names, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
 * (tuple of ints size, Number fill_value, *, Tensor out, torch.dtype dtype, torch.layout layout, torch.device device, bool pin_memory, bool requires_grad)
```

I assumed you meant to enclose it within a list or tuple, as in the declaration of `shape`: `shape = [dim.get_dim_value() for dim in dims]`. I did that, and I got the following warning:

```
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:676: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  shape = (torch.tensor(shape, dtype=torch.long),)
```

Either list or tuple yielded the same error. I took the liberty to perform an explicit casting:

```
```

And it looks like the code is working! See the [ONNX exporting tool's output](https://gist.github.com/Icemole/9a2aea10fb01304ebd3f74d95489682e) and the associated graph. I'll test this soon, but right now I'll push the fix.

**Comment by albertz** [comment]
> I assumed you meant to enclose it within a list or tuple, as in the declaration of `shape`: `shape = [dim.get_dim_value() for dim in dims]`

No, I meant it in the way I wrote it. I did not know that this is not supported. But then it should be:
```python
if torch.onnx.is_in_onnx_export():
    # onnx::ConstantOfShape (via torch.full) must get shape as int64.
    # https://github.com/rwth-i6/returnn/issues/1333#issuecomment-1607236783
    shape = [torch.tensor(dim, dtype=torch.long) for dim in shape]
```
Btw, your code is a bit buggy. You cannot call `long()` as the dim value might also just be an `int` object, not necessarily a `torch.Tensor`.

>  I took the liberty to perform an explicit casting

I'm a bit confused now. I thought this is what we were anyway talking about? What did you mean before if not that?

**Comment by albertz** [comment]
> > `out = torch.full([torch.tensor(3, dtype=torch.int32)], 0)`
> 
> This standalone code doesn't fail.

What do you mean? It should reproduce the warning (error?) or not? If not, then you did sth wrong? Or what exactly in this code is different from our case?

Did you report the error? Of course using a minimal code example. Like this code.

**Comment by albertz** [comment]
> And it looks like the code is working! See the [ONNX exporting tool's output](https://gist.github.com/Icemole/9a2aea10fb01304ebd3f74d95489682e) and the associated graph.

Nice.

We should extend the test case to also perform the inference, similar as in the other test code I send before.

We also should check those warnings:
```
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/tensor/_tensor_extra.py:327: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if raw_shape[i] != self.batch_shape[i]:
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/frontend/_backend.py:136: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert all(dim is None or dim == existing_shape[i] for i, dim in enumerate(shape))
```
Maybe it's fine but we should double check.

**Comment by Icemole** [comment]
> `shape = [torch.tensor(dim, dtype=torch.long) for dim in shape]`

As per the error code I shared before,

> [...] torch.tensor results are registered as constants in the trace. [...]

using `torch.tensor()` to create a tensor hardcodes that tensor every time that call is done. Therefore, we'd be getting always `torch.tensor([16, 16, 16, 16, 16])`, no matter what the input sizes are, if I understood the warning correctly. This is why I casted the tensor with `long()` instead of directly using `torch.tensor()`.

**Comment by Icemole** [comment]
> > I took the liberty to perform an explicit casting
>
> I'm a bit confused now. I thought this is what we were anyway talking about? What did you mean before if not that?

Yes, my wording was wrong. I meant an alternative way to perform this casting.

---

> > > out = torch.full([torch.tensor(3, dtype=torch.int32)], 0)
> >
> > This standalone code doesn't fail.
>
> It should reproduce the warning (error?) or not?

If you mean substituting the call to `raw_tensor = rf.full(...)` with `raw_tensor = torch.full([torch.tensor(3, dtype=torch.int32)], 0, dtype=torch.int32)`, then the code does not yield the error we were originally concerned about, although it warns the same as your code that uses `torch.tensor()` for the explicit casting:

```
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:677: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  raw_tensor = torch.full([torch.tensor(3, dtype=torch.int32)], 0, dtype=torch.int32)
```

If, on the other hand, you simply mean doing this in a python console:

```
>>> import torch
>>> out = torch.full([torch.tensor(3, dtype=torch.int32)], 0)
```

Then this code does not reproduce the error.

**Comment by albertz** [comment]
I'm speaking about this warning (error), nothing else here:
```
/home/nbeneitez/.venv/returnn_pytorch/lib/python3.10/site-packages/torch/onnx/utils.py:1636:
UserWarning: The exported ONNX model failed ONNX shape inference.The model will not be executable by the
ONNX Runtime.If this is unintended and you believe there is a bug,please report an issue at
https://github.com/pytorch/pytorch/issues.Error reported by strict ONNX shape inference:
[ShapeInferenceError] (op_type:ConstantOfShape, node name: ConstantOfShape_14): input typestr: T1,
has unsupported type: tensor(int32) (Triggered internally at ../torch/csrc/jit/serialization/export.cpp:1407.)
  _C._check_onnx_proto(proto)
```

I thought the relevant part is that the shape is in int32, or not? So what minimal code, standalone code (not using RETURNN at all), assumably using `torch.full`, would reproduce this?

**Comment by albertz** [comment]
Regarding getting rid of the `TracerWarning`s for those sanity checking code, we could use such code:
```python
with warnings.catch_warnings():
    # PyTorch JIT tracing will produce this warning:
    # TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect.
    #   We can't record the data flow of Python values,
    #   so this value will be treated as a constant in the future.
    #   This means that the trace might not generalize to other inputs!
    # This is fine for us, that we do this check only once, and then not at later runtime anymore.
    warnings.filterwarnings("ignore", message="")
    # <now the original code which produced the warning>
```
However I'm not sure if this is the best way. Also, this should be encapsulated somehow in a nicer way, as the triggering code is actually not PyTorch specific.

Or we just ignore it for now.

**Comment by Icemole** [comment]
As a completely different note, a constant value in the constructor of `torch.tensor()` yields two warnings that are different from the one described above:

```
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:676: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  shape = [torch.tensor(dim, dtype=torch.int64) for dim in shape]
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:676: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).
  shape = [torch.tensor(dim, dtype=torch.int64) for dim in shape]
```

This is the warning I'm worried about with the current code that you fixed. I understand that my previous implementation was bugged because of the reasons you mentioned, but with the creation of a tensor we're adding another two warnings. I'm worried by the fact that, since the call to `torch.tensor()` returns some `tensor(X)`, such tensor would be registered in the graph as a constant. The reference to the code you fixed in the graph is the following:
```
  %onnx::Unsqueeze_23 : Long(requires_grad=0, device=cpu) = onnx::Cast[to=7](%classes:size0) # /home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/torch/frontend/_backend.py:676:0
```

It doesn't seem like a constant to me, but maybe we could simply get rid of the warning by casting with `long()` as I did, but this time checking if the object to cast was a tensor. I'm not sure if the call to `sourceTensor.clone().detach()` (as the second warning suggests) can cast the tensor to any other type.

**Comment by albertz** [comment]
> > So what minimal code, standalone code (not using RETURNN at all), assumably using torch.full, would reproduce this?
> 
> I understood now. [This setup](https://gist.github.com/Icemole/38e03bc2bb0e96d8bc3423b7d21511aa) fails with the expected warning that you say.

So this looks like a bug in the ONNX export of PyTorch, right? Because running the PyTorch code as-is works completely fine.

Can you make a bug report and link it here?

**Comment by albertz** [comment]
> As a completely different note, a constant value in the constructor of `torch.tensor()` yields two warnings that are different from the one described above ...

Yes I changed the code again. I thought that `torch.tensor` would work but I was wrong. In the current code, there should be no such warning, right?

**Comment by Icemole** [comment]
> So this looks like a bug in the ONNX export of PyTorch, right? Because running the PyTorch code as-is works completely fine.

You mean that it's a bug because the operation is converted to the same `onnx::Unsqueeze` operation regardless of whether the tensor is being created again with `torch.tensor()` or being explicitly casted with `tensor.long()`, but in the first approach it yields a warning and in the second approach it doesn't, right?

> In the current code, there should be no such warning, right?

Indeed, there's no such warning :+1: the only warnings are the following ones, which we already knew of their existence:

```
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/tensor/_tensor_extra.py:327: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if raw_shape[i] != self.batch_shape[i]:
/home/nbeneitez/Documentos/work/repos/returnn_pytorch/returnn/frontend/_backend.py:136: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  assert all(dim is None or dim == existing_shape[i] for i, dim in enumerate(shape))
WARNING:root:Output 'output' Tensor{'raw_tensor', [B?,'time:var-unk:output'[B?],F'feature:output'(2)]}: Cannot infer dynamic size for dim Dim{'time:var-unk:output'[B?]}. Using Tensor{'full', [B?], dtype='int32'} as fallback.
```

**Comment by Icemole** [comment]
I was able to create a [better minimal example](https://gist.github.com/Icemole/47dc9757678ed5bff36cc222627525de) detailing each of the three cases:

- Not converting from `int32`. This results in the error with which we're already familiar: `UserWarning: The exported ONNX model failed ONNX shape inference.`. Final graph [here](https://gist.github.com/Icemole/3eedac24a108f63cb441defb7897d039).
- Converting from `int32` to `int64` through `torch.tensor()`. This results in a warning that I described in my last message: `TracerWarning: torch.tensor results are registered as constants in the trace.` Final graph [here](https://gist.github.com/Icemole/c4e75dacbc8593a42208835940cd1fb6).
- Converting from `int32` to `int64` through `tensor.long()`. This doesn't result in a warning at all, and the graph is the same as the option above. Final graph [here](https://gist.github.com/Icemole/3617ba90038a4e8af6efe116d06b4651).

Each of the three gives different warnings, but as I briefly stated above, converting by both ways yields the same graph. Not converting doesn't yield the same graph with respect to the other two, but I think the only difference is the lack of `onnx::Cast` operators and the final type of the shape (`Int` in the case of not converting, `Long` in the case of converting). Now there's no `onnx::ConstantOfShape` operator though...

**Comment by albertz** [comment]
> > So this looks like a bug in the ONNX export of PyTorch, right? Because running the PyTorch code as-is works completely fine.
> 
> You mean that it's a bug because the operation is converted to the same `onnx::Unsqueeze` operation regardless of whether the tensor is being created again with `torch.tensor()` or being explicitly casted with `tensor.long()`, but in the first approach it yields a warning and in the second approach it doesn't, right?

No. It's a bug simply because the PyTorch works without ONNX, but then fails in ONNX export. No matter the details, that already can be considered a bug (or maybe some missing functionality). You said that your example code ([this](https://gist.github.com/Icemole/38e03bc2bb0e96d8bc3423b7d21511aa)) shows this.

You should not mix different aspects in the bug report, like converting int32 to int64 or so. Keep separate issues as separate issues.

In the issue, the example code is one of the most important aspects. You directly should show the code additionally to linking the Gist.

In the issue, you formatted the error in a way that you cannot easily read it. I think somehow you removed the newlines from it? It would be helpful to add back the line breaks such that it can be read more easily.

In the issue, you now made the example code more complex than it needs to be. But such example code should always be as minimal as possible. Why do you need those `d1`, `d2`, `d3`? Why not just one dimension, as in the code you had before? Also, you are now mixing three different things in there, which could be a bit confusing. That option 2 what you mention there is anyway totally clear. It's not at all a bug, so it doesn't really belong there. It was our own mistake, that we understood wrong how `torch.tensor(...)` works. Nothing is wrong with that.

**Comment by Icemole** [comment]
I will reformat the issue as discussed.

**Comment by albertz** [comment]
Thanks. But as you wrote it, it is still a bit unclear to me. Are there two types of bugs, for option 1 and option 2? Then that should be two separate bug reports. Or are they the same bug? Or is there actually no bug with option 1? Then why is option 1 relevant for the bug report?

> When converting from int32 to int64 through tensor_object.long() (final graph [here](https://gist.github.com/Icemole/3617ba90038a4e8af6efe116d06b4651)), we don't get any warning.

So, this is option 1? So with option 1, all works as intended?

Then you should not put option 1 even into the code. The code should demonstrate the bug. I think this is confusing. You can mention in a remark that it works correctly when you call `.long()` first.

Btw, you forgot to enable Python syntax highlighting for the code. It's hard to read. I cannot easily see what part is the commented out code. But anyway, you better should not post any code where there is commented out code.

Btw, `shape = [dim.long() for dim in shape]` is also too complicated. Just write `shape = [d1.long()]`, to see the difference. But as said, I think you should not put this into the code at all, but just leave it as a comment.

**Comment by Icemole** [comment]
I've formatted the issue as specified.

Moving on to another topic, the inference seems to be working, or at least it's producing some reasonable output from the random input I gave it. However, I get two warnings:
```python
2023-06-30 13:33:33.516802915 [W:onnxruntime:, execution_frame.cc:857 VerifyOutputSizes] Expected shape from model of {} does not match actual shape of {3} for output output:size0
2023-06-30 13:33:33.517066362 [W:onnxruntime:, execution_frame.cc:857 VerifyOutputSizes] Expected shape from model of {-1} does not match actual shape of {3,50,2} for output output:size1
```

As I understand, the tool is failing to get the output shapes. I'm running the ONNX exporting tool with the torch demo. The code I'm running for inference is the following:
```python
import onnxruntime as ort
import torch


onnx_model = "/tmp/nbeneitez/returnn/demos/demo-torch/model.005.onnx"
torch.manual_seed(42)
dummy_data = torch.randn([3, 50, 9])
dummy_data_len = torch.IntTensor([3, 50, 1])

session = ort.InferenceSession(onnx_model)
outputs_onnx = session.run(None, {"data": dummy_data.numpy(), "classes:size0": dummy_data_len.numpy()})

print(torch.FloatTensor(outputs_onnx[0]))
print(torch.IntTensor(outputs_onnx[1]))
print(torch.IntTensor(outputs_onnx[2]))
```

I adapted the inference code from [this code](https://github.com/rwth-i6/i6_models/blob/52410fafa72b20cee174ece96d5d850bc73c4c61/tests/test_blstm.py#L36).

**Comment by Icemole** [comment]
Thanks for the heads-up.

I tried inserting the batch size and the sequence lengths in `classes:size0` and `classes:size1` as you requested, but my inference code failed with the following error:
```
  File "/home/nbeneitez/.venv/returnn_pytorch/lib/python3.10/site-packages/onnxruntime
/capi/onnxruntime_inference_collection.py", line 217, in run
    return self._sess.run(output_names, input_feed, run_options)
onnxruntime.capi.onnxruntime_pybind11_state.InvalidArgument: [ONNXRuntimeError] : 2 : INVALID_ARGUMENT :
Invalid Feed Input Name:classes:size1
```

Indeed, in the [graph](https://gist.github.com/Icemole/9a2aea10fb01304ebd3f74d95489682e) we obtained when the tool initially worked (original comment [here](https://github.com/rwth-i6/returnn/issues/1333#issuecomment-1607474273)), the only two inputs I can see are `data` and `classes:size0`. I assume the rest of the inputs aren't needed, but they're properly specified in the input names provided to `torch.onnx.export`. Even if they might not be used, do you think they should be disregarded?

A better view of the graph through through [NETRON](https://netron.app/):
![model 005 onnx](https://github.com/rwth-i6/returnn/assets/31628502/580737b5-8b72-43b3-aadb-694e1c0c3cf8)

**Comment by albertz** [comment]
> Even if they might not be used, do you think they should be disregarded as inputs to the final graph, when we explicitly stated that these would be inputs to the graph?

Who do you ask here? You say this is maybe unexpected behavior of the PyTorch ONNX export tool? Yes, maybe. Did you check whether this is discussed anywhere? If not, can you maybe open an issue on this, and ask if this is a bug or expected behavior?

Maybe the answer is, this is expected behavior. Or maybe it would be fixed in a future version but we probably also want to make it work for the current version.

So then the answer goes back to us. What do *we*  actually want to do in such case?

I think we want that we always have a clear interface. When the interface is that we give it (input, in_seq_lens), we want that the resulting ONNX graph can always get this type of input, no matter if it is actually used or not.

Btw, via this argument, we might also say: The interface is that the output should be (output, out_seq_lens). And out_seq_lens should be properly defined. So then our current heuristic to just fill it with dummy values is not good, and we want that the user makes it explicit?

We should maybe discuss with @curufinwe about this. It matters for RASR, what type of interface to expect there. Also when RASR wants to do batched recognition at some point, and when RASR wants to support different sequence lengths properly, with potential downsampling in the model, the in_seq_lens and out_seq_lens are both important.

Btw, the batch_dim in the input is actually sth which we might want to remove, as it is redundant. In general, any dims which have a scalar dyn_size_ext are not needed. We probably should change this.

**Comment by Icemole** [comment]
I pushed tests for inference to a new branch, please let me know if something needs to be changed.

I don't have more time for the week to work in the ONNX exporting but I quickly went through your comment and overall I agree, we should bring up this up with more people. And yes, I would rather have a constant interface always, even if the inputs end up calculating nothing, since that can be better automatically handled. I will try to examine some issues or details further next week.

**Comment by albertz** [comment]
> I pushed tests for inference to a new branch, please let me know if something needs to be changed.

I don't remember anymore, what was this about? Or just about extending our tests by actually forwarding sth through the ONNX graph via `onnxruntime`?

Why did you rename `_test_torch_export_to_onnx` to `_test_export_to_onnx`? The old name `_test_torch_export_to_onnx` sounds more explicit and better to me.

Why is there no PR for this branch?

**Comment by Icemole** [comment]
> [...] what was this about? [...]

As you said, this was just about extending our tests by actually forwarding something through the ONNX graph via `onnxruntime`. We wanted to have some tests that checked not only the graph creation but also the inference, so I made these.

> Why did you rename [...]

I understand that `torch` there meant the technology itself, but it was a little confusing for me because I initially thought `torch` was the demo name, and this code worked for both the `torch` and the `rf` demos. I can bring it back if you think it was better.

> Why is there no PR [...]

I've been focused on other tasks and didn't have time to invest on the ONNX tool. I'll address the change above and create the PR, but I can't do much more.

**Comment by albertz** [comment]
Last week, we decided:

ONNX input/output consistency, three use cases for RASR:
1. Only inputs -> only outputs
2. Inputs + seq lens -> outputs: assume output seq lens == input seq lens
  a. This is what's currently specified at the moment (seq lens are optional, so actually 1 and 2 are supported)
3. Inputs + seq lens -> outputs + seq lens: assume output seq lens are correct (don't randomly fill), we should be able to use them later on

User should explicitly specify each of the cases

**Comment by albertz** [comment]
Btw, I wonder, the resulting ONNX graph, this is a freezed graph, so it contains all the parameters? (@Icemole)

**Comment by albertz** [comment]
I think we can close this. In case there is any issue, please open a new issue (maybe refer to this one here).
